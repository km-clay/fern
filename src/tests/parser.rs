use super::super::*;

#[test]
fn parse_simple() {
	let input = "echo hello world";
	let tk_stream: Vec<_> = LexStream::new(Rc::new(input.to_string()), LexFlags::empty())
		.map(|tk| tk.unwrap())
		.collect();
	let nodes: Vec<_> = ParseStream::new(tk_stream).collect();

	insta::assert_debug_snapshot!(nodes)
}

#[test]
fn parse_pipeline() {
	let input = "echo foo | sed s/foo/bar";
	let tk_stream: Vec<_> = LexStream::new(Rc::new(input.to_string()), LexFlags::empty())
		.map(|tk| tk.unwrap())
		.collect();
	let nodes: Vec<_> = ParseStream::new(tk_stream).collect();

	insta::assert_debug_snapshot!(nodes)
}

#[test]
fn parse_conjunction() {
	let input = "echo foo && echo bar";
	let tk_stream: Vec<_> = LexStream::new(Rc::new(input.to_string()), LexFlags::empty())
		.map(|tk| tk.unwrap())
		.collect();
	let nodes: Vec<_> = ParseStream::new(tk_stream).collect();

	insta::assert_debug_snapshot!(nodes)
}

#[test]
fn parse_conjunction_and_pipeline() {
	let input = "echo foo | sed s/foo/bar/ && echo bar | sed s/bar/foo/ || echo foo bar | sed s/foo bar/bar foo/";
	let tk_stream: Vec<_> = LexStream::new(Rc::new(input.to_string()), LexFlags::empty())
		.map(|tk| tk.unwrap())
		.collect();
	let nodes: Vec<_> = ParseStream::new(tk_stream).collect();

	insta::assert_debug_snapshot!(nodes)
}

#[test]
fn parse_multiline() {
	let input = "
echo hello world
echo foo bar
echo boo biz";
	let tk_stream: Vec<_> = LexStream::new(Rc::new(input.to_string()), LexFlags::empty())
		.map(|tk| tk.unwrap())
		.collect();
	let nodes: Vec<_> = ParseStream::new(tk_stream).collect();

	insta::assert_debug_snapshot!(nodes)
}

#[test]
fn parse_if_simple() {
	let input = "if foo; then echo bar; fi";
	let tk_stream: Vec<_> = LexStream::new(Rc::new(input.to_string()), LexFlags::empty())
		.map(|tk| tk.unwrap())
		.collect();
	let nodes: Vec<_> = ParseStream::new(tk_stream).collect();

	insta::assert_debug_snapshot!(nodes)
}
#[test]
fn parse_if_with_elif() {
	let input = "if foo; then echo bar; elif bar; then echo foo; fi";
	let tk_stream: Vec<_> = LexStream::new(Rc::new(input.to_string()), LexFlags::empty())
		.map(|tk| tk.unwrap())
		.collect();
	let nodes: Vec<_> = ParseStream::new(tk_stream).collect();

	insta::assert_debug_snapshot!(nodes)
}
#[test]
fn parse_if_multiple_elif() {
	let input = "if foo; then echo bar; elif bar; then echo foo; elif biz; then echo baz; fi";
	let tk_stream: Vec<_> = LexStream::new(Rc::new(input.to_string()), LexFlags::empty())
		.map(|tk| tk.unwrap())
		.collect();
	let nodes: Vec<_> = ParseStream::new(tk_stream).collect();

	insta::assert_debug_snapshot!(nodes)
}
#[test]
fn parse_if_multiline() {
	let input = "
if foo; then
	echo bar
elif bar; then
	echo foo;
elif biz; then
	echo baz
fi";
	let tk_stream: Vec<_> = LexStream::new(Rc::new(input.to_string()), LexFlags::empty())
		.map(|tk| tk.unwrap())
		.collect();
	let nodes: Vec<_> = ParseStream::new(tk_stream).collect();

	insta::assert_debug_snapshot!(nodes)
}
#[test]
fn parse_loop_simple() {
	let input = "while foo; do bar; done";
	let tk_stream: Vec<_> = LexStream::new(Rc::new(input.to_string()), LexFlags::empty())
		.map(|tk| tk.unwrap())
		.collect();
	let nodes: Vec<_> = ParseStream::new(tk_stream).collect();

	insta::assert_debug_snapshot!(nodes)
}
#[test]
fn parse_loop_until() {
	let input = "until foo; do bar; done";
	let tk_stream: Vec<_> = LexStream::new(Rc::new(input.to_string()), LexFlags::empty())
		.map(|tk| tk.unwrap())
		.collect();
	let nodes: Vec<_> = ParseStream::new(tk_stream).collect();

	insta::assert_debug_snapshot!(nodes)
}
#[test]
fn parse_loop_multiline() {
	let input = "
until foo; do
	bar
done";
	let tk_stream: Vec<_> = LexStream::new(Rc::new(input.to_string()), LexFlags::empty())
		.map(|tk| tk.unwrap())
		.collect();
	let nodes: Vec<_> = ParseStream::new(tk_stream).collect();

	insta::assert_debug_snapshot!(nodes)
}
#[test]
fn parse_case_simple() {
	let input = "case foo in foo) bar;; bar) foo;; biz) baz;; esac";
	let tk_stream: Vec<_> = LexStream::new(Rc::new(input.to_string()), LexFlags::empty())
		.map(|tk| tk.unwrap())
		.collect();
	let nodes: Vec<_> = ParseStream::new(tk_stream).collect();

	insta::assert_debug_snapshot!(nodes)
}
#[test]
fn parse_case_multiline() {
	let input = "case foo in
	foo) bar
	;;
	bar) foo
	;;
	biz) baz
	;;
esac";
	let tk_stream: Vec<_> = LexStream::new(Rc::new(input.to_string()), LexFlags::empty())
		.map(|tk| tk.unwrap())
		.collect();
	let nodes: Vec<_> = ParseStream::new(tk_stream).collect();

	insta::assert_debug_snapshot!(nodes)
}
#[test]
fn parse_case_nested() {
	let input = "case foo in
	foo) if true; then
		echo foo
	fi
	;;
	bar) if false; then
		echo bar
	fi
	;;
esac";
	let tk_stream: Vec<_> = LexStream::new(Rc::new(input.to_string()), LexFlags::empty())
		.map(|tk| tk.unwrap())
		.collect();
	let nodes: Vec<_> = ParseStream::new(tk_stream).collect();

	insta::assert_debug_snapshot!(nodes)
}
