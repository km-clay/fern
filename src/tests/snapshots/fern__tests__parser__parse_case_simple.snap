---
source: src/tests/parser.rs
expression: nodes
---
[
    Ok(
        Node {
            class: Conjunction {
                elements: [
                    ConjunctNode {
                        cmd: Node {
                            class: CaseNode {
                                pattern: Tk {
                                    class: Str,
                                    span: Span {
                                        range: 5..8,
                                        source: "case foo in foo) bar;; bar) foo;; biz) baz;; esac",
                                    },
                                    flags: TkFlags(
                                        0x0,
                                    ),
                                },
                                case_blocks: [
                                    CaseNode {
                                        pattern: Tk {
                                            class: CasePattern,
                                            span: Span {
                                                range: 12..16,
                                                source: "case foo in foo) bar;; bar) foo;; biz) baz;; esac",
                                            },
                                            flags: TkFlags(
                                                0x0,
                                            ),
                                        },
                                        body: [
                                            Node {
                                                class: Pipeline {
                                                    cmds: [
                                                        Node {
                                                            class: Command {
                                                                assignments: [],
                                                                argv: [
                                                                    Tk {
                                                                        class: Str,
                                                                        span: Span {
                                                                            range: 17..20,
                                                                            source: "case foo in foo) bar;; bar) foo;; biz) baz;; esac",
                                                                        },
                                                                        flags: TkFlags(
                                                                            IS_CMD,
                                                                        ),
                                                                    },
                                                                ],
                                                            },
                                                            flags: NdFlags(
                                                                0x0,
                                                            ),
                                                            redirs: [],
                                                            tokens: [
                                                                Tk {
                                                                    class: Str,
                                                                    span: Span {
                                                                        range: 17..20,
                                                                        source: "case foo in foo) bar;; bar) foo;; biz) baz;; esac",
                                                                    },
                                                                    flags: TkFlags(
                                                                        IS_CMD,
                                                                    ),
                                                                },
                                                                Tk {
                                                                    class: Sep,
                                                                    span: Span {
                                                                        range: 20..23,
                                                                        source: "case foo in foo) bar;; bar) foo;; biz) baz;; esac",
                                                                    },
                                                                    flags: TkFlags(
                                                                        0x0,
                                                                    ),
                                                                },
                                                            ],
                                                        },
                                                    ],
                                                    pipe_err: false,
                                                },
                                                flags: NdFlags(
                                                    0x0,
                                                ),
                                                redirs: [],
                                                tokens: [
                                                    Tk {
                                                        class: Str,
                                                        span: Span {
                                                            range: 17..20,
                                                            source: "case foo in foo) bar;; bar) foo;; biz) baz;; esac",
                                                        },
                                                        flags: TkFlags(
                                                            IS_CMD,
                                                        ),
                                                    },
                                                    Tk {
                                                        class: Sep,
                                                        span: Span {
                                                            range: 20..23,
                                                            source: "case foo in foo) bar;; bar) foo;; biz) baz;; esac",
                                                        },
                                                        flags: TkFlags(
                                                            0x0,
                                                        ),
                                                    },
                                                ],
                                            },
                                        ],
                                    },
                                    CaseNode {
                                        pattern: Tk {
                                            class: CasePattern,
                                            span: Span {
                                                range: 23..27,
                                                source: "case foo in foo) bar;; bar) foo;; biz) baz;; esac",
                                            },
                                            flags: TkFlags(
                                                0x0,
                                            ),
                                        },
                                        body: [
                                            Node {
                                                class: Pipeline {
                                                    cmds: [
                                                        Node {
                                                            class: Command {
                                                                assignments: [],
                                                                argv: [
                                                                    Tk {
                                                                        class: Str,
                                                                        span: Span {
                                                                            range: 28..31,
                                                                            source: "case foo in foo) bar;; bar) foo;; biz) baz;; esac",
                                                                        },
                                                                        flags: TkFlags(
                                                                            IS_CMD,
                                                                        ),
                                                                    },
                                                                ],
                                                            },
                                                            flags: NdFlags(
                                                                0x0,
                                                            ),
                                                            redirs: [],
                                                            tokens: [
                                                                Tk {
                                                                    class: Str,
                                                                    span: Span {
                                                                        range: 28..31,
                                                                        source: "case foo in foo) bar;; bar) foo;; biz) baz;; esac",
                                                                    },
                                                                    flags: TkFlags(
                                                                        IS_CMD,
                                                                    ),
                                                                },
                                                                Tk {
                                                                    class: Sep,
                                                                    span: Span {
                                                                        range: 31..34,
                                                                        source: "case foo in foo) bar;; bar) foo;; biz) baz;; esac",
                                                                    },
                                                                    flags: TkFlags(
                                                                        0x0,
                                                                    ),
                                                                },
                                                            ],
                                                        },
                                                    ],
                                                    pipe_err: false,
                                                },
                                                flags: NdFlags(
                                                    0x0,
                                                ),
                                                redirs: [],
                                                tokens: [
                                                    Tk {
                                                        class: Str,
                                                        span: Span {
                                                            range: 28..31,
                                                            source: "case foo in foo) bar;; bar) foo;; biz) baz;; esac",
                                                        },
                                                        flags: TkFlags(
                                                            IS_CMD,
                                                        ),
                                                    },
                                                    Tk {
                                                        class: Sep,
                                                        span: Span {
                                                            range: 31..34,
                                                            source: "case foo in foo) bar;; bar) foo;; biz) baz;; esac",
                                                        },
                                                        flags: TkFlags(
                                                            0x0,
                                                        ),
                                                    },
                                                ],
                                            },
                                        ],
                                    },
                                    CaseNode {
                                        pattern: Tk {
                                            class: CasePattern,
                                            span: Span {
                                                range: 34..38,
                                                source: "case foo in foo) bar;; bar) foo;; biz) baz;; esac",
                                            },
                                            flags: TkFlags(
                                                0x0,
                                            ),
                                        },
                                        body: [
                                            Node {
                                                class: Pipeline {
                                                    cmds: [
                                                        Node {
                                                            class: Command {
                                                                assignments: [],
                                                                argv: [
                                                                    Tk {
                                                                        class: Str,
                                                                        span: Span {
                                                                            range: 39..42,
                                                                            source: "case foo in foo) bar;; bar) foo;; biz) baz;; esac",
                                                                        },
                                                                        flags: TkFlags(
                                                                            IS_CMD,
                                                                        ),
                                                                    },
                                                                ],
                                                            },
                                                            flags: NdFlags(
                                                                0x0,
                                                            ),
                                                            redirs: [],
                                                            tokens: [
                                                                Tk {
                                                                    class: Str,
                                                                    span: Span {
                                                                        range: 39..42,
                                                                        source: "case foo in foo) bar;; bar) foo;; biz) baz;; esac",
                                                                    },
                                                                    flags: TkFlags(
                                                                        IS_CMD,
                                                                    ),
                                                                },
                                                                Tk {
                                                                    class: Sep,
                                                                    span: Span {
                                                                        range: 42..45,
                                                                        source: "case foo in foo) bar;; bar) foo;; biz) baz;; esac",
                                                                    },
                                                                    flags: TkFlags(
                                                                        0x0,
                                                                    ),
                                                                },
                                                            ],
                                                        },
                                                    ],
                                                    pipe_err: false,
                                                },
                                                flags: NdFlags(
                                                    0x0,
                                                ),
                                                redirs: [],
                                                tokens: [
                                                    Tk {
                                                        class: Str,
                                                        span: Span {
                                                            range: 39..42,
                                                            source: "case foo in foo) bar;; bar) foo;; biz) baz;; esac",
                                                        },
                                                        flags: TkFlags(
                                                            IS_CMD,
                                                        ),
                                                    },
                                                    Tk {
                                                        class: Sep,
                                                        span: Span {
                                                            range: 42..45,
                                                            source: "case foo in foo) bar;; bar) foo;; biz) baz;; esac",
                                                        },
                                                        flags: TkFlags(
                                                            0x0,
                                                        ),
                                                    },
                                                ],
                                            },
                                        ],
                                    },
                                ],
                            },
                            flags: NdFlags(
                                0x0,
                            ),
                            redirs: [],
                            tokens: [
                                Tk {
                                    class: Str,
                                    span: Span {
                                        range: 0..4,
                                        source: "case foo in foo) bar;; bar) foo;; biz) baz;; esac",
                                    },
                                    flags: TkFlags(
                                        KEYWORD,
                                    ),
                                },
                                Tk {
                                    class: Str,
                                    span: Span {
                                        range: 5..8,
                                        source: "case foo in foo) bar;; bar) foo;; biz) baz;; esac",
                                    },
                                    flags: TkFlags(
                                        0x0,
                                    ),
                                },
                                Tk {
                                    class: Str,
                                    span: Span {
                                        range: 9..11,
                                        source: "case foo in foo) bar;; bar) foo;; biz) baz;; esac",
                                    },
                                    flags: TkFlags(
                                        KEYWORD,
                                    ),
                                },
                                Tk {
                                    class: CasePattern,
                                    span: Span {
                                        range: 12..16,
                                        source: "case foo in foo) bar;; bar) foo;; biz) baz;; esac",
                                    },
                                    flags: TkFlags(
                                        0x0,
                                    ),
                                },
                                Tk {
                                    class: Str,
                                    span: Span {
                                        range: 17..20,
                                        source: "case foo in foo) bar;; bar) foo;; biz) baz;; esac",
                                    },
                                    flags: TkFlags(
                                        IS_CMD,
                                    ),
                                },
                                Tk {
                                    class: Sep,
                                    span: Span {
                                        range: 20..23,
                                        source: "case foo in foo) bar;; bar) foo;; biz) baz;; esac",
                                    },
                                    flags: TkFlags(
                                        0x0,
                                    ),
                                },
                                Tk {
                                    class: CasePattern,
                                    span: Span {
                                        range: 23..27,
                                        source: "case foo in foo) bar;; bar) foo;; biz) baz;; esac",
                                    },
                                    flags: TkFlags(
                                        0x0,
                                    ),
                                },
                                Tk {
                                    class: Str,
                                    span: Span {
                                        range: 28..31,
                                        source: "case foo in foo) bar;; bar) foo;; biz) baz;; esac",
                                    },
                                    flags: TkFlags(
                                        IS_CMD,
                                    ),
                                },
                                Tk {
                                    class: Sep,
                                    span: Span {
                                        range: 31..34,
                                        source: "case foo in foo) bar;; bar) foo;; biz) baz;; esac",
                                    },
                                    flags: TkFlags(
                                        0x0,
                                    ),
                                },
                                Tk {
                                    class: CasePattern,
                                    span: Span {
                                        range: 34..38,
                                        source: "case foo in foo) bar;; bar) foo;; biz) baz;; esac",
                                    },
                                    flags: TkFlags(
                                        0x0,
                                    ),
                                },
                                Tk {
                                    class: Str,
                                    span: Span {
                                        range: 39..42,
                                        source: "case foo in foo) bar;; bar) foo;; biz) baz;; esac",
                                    },
                                    flags: TkFlags(
                                        IS_CMD,
                                    ),
                                },
                                Tk {
                                    class: Sep,
                                    span: Span {
                                        range: 42..45,
                                        source: "case foo in foo) bar;; bar) foo;; biz) baz;; esac",
                                    },
                                    flags: TkFlags(
                                        0x0,
                                    ),
                                },
                                Tk {
                                    class: Str,
                                    span: Span {
                                        range: 45..49,
                                        source: "case foo in foo) bar;; bar) foo;; biz) baz;; esac",
                                    },
                                    flags: TkFlags(
                                        KEYWORD,
                                    ),
                                },
                            ],
                        },
                        operator: Null,
                    },
                ],
            },
            flags: NdFlags(
                0x0,
            ),
            redirs: [],
            tokens: [
                Tk {
                    class: Str,
                    span: Span {
                        range: 0..4,
                        source: "case foo in foo) bar;; bar) foo;; biz) baz;; esac",
                    },
                    flags: TkFlags(
                        KEYWORD,
                    ),
                },
                Tk {
                    class: Str,
                    span: Span {
                        range: 5..8,
                        source: "case foo in foo) bar;; bar) foo;; biz) baz;; esac",
                    },
                    flags: TkFlags(
                        0x0,
                    ),
                },
                Tk {
                    class: Str,
                    span: Span {
                        range: 9..11,
                        source: "case foo in foo) bar;; bar) foo;; biz) baz;; esac",
                    },
                    flags: TkFlags(
                        KEYWORD,
                    ),
                },
                Tk {
                    class: CasePattern,
                    span: Span {
                        range: 12..16,
                        source: "case foo in foo) bar;; bar) foo;; biz) baz;; esac",
                    },
                    flags: TkFlags(
                        0x0,
                    ),
                },
                Tk {
                    class: Str,
                    span: Span {
                        range: 17..20,
                        source: "case foo in foo) bar;; bar) foo;; biz) baz;; esac",
                    },
                    flags: TkFlags(
                        IS_CMD,
                    ),
                },
                Tk {
                    class: Sep,
                    span: Span {
                        range: 20..23,
                        source: "case foo in foo) bar;; bar) foo;; biz) baz;; esac",
                    },
                    flags: TkFlags(
                        0x0,
                    ),
                },
                Tk {
                    class: CasePattern,
                    span: Span {
                        range: 23..27,
                        source: "case foo in foo) bar;; bar) foo;; biz) baz;; esac",
                    },
                    flags: TkFlags(
                        0x0,
                    ),
                },
                Tk {
                    class: Str,
                    span: Span {
                        range: 28..31,
                        source: "case foo in foo) bar;; bar) foo;; biz) baz;; esac",
                    },
                    flags: TkFlags(
                        IS_CMD,
                    ),
                },
                Tk {
                    class: Sep,
                    span: Span {
                        range: 31..34,
                        source: "case foo in foo) bar;; bar) foo;; biz) baz;; esac",
                    },
                    flags: TkFlags(
                        0x0,
                    ),
                },
                Tk {
                    class: CasePattern,
                    span: Span {
                        range: 34..38,
                        source: "case foo in foo) bar;; bar) foo;; biz) baz;; esac",
                    },
                    flags: TkFlags(
                        0x0,
                    ),
                },
                Tk {
                    class: Str,
                    span: Span {
                        range: 39..42,
                        source: "case foo in foo) bar;; bar) foo;; biz) baz;; esac",
                    },
                    flags: TkFlags(
                        IS_CMD,
                    ),
                },
                Tk {
                    class: Sep,
                    span: Span {
                        range: 42..45,
                        source: "case foo in foo) bar;; bar) foo;; biz) baz;; esac",
                    },
                    flags: TkFlags(
                        0x0,
                    ),
                },
                Tk {
                    class: Str,
                    span: Span {
                        range: 45..49,
                        source: "case foo in foo) bar;; bar) foo;; biz) baz;; esac",
                    },
                    flags: TkFlags(
                        KEYWORD,
                    ),
                },
            ],
        },
    ),
]
